File: problem/Sum_Root_Leaf/sumRootToLeaf.js
var sumRootToLeaf = function(root) {
    const dfs = (node, curr) => {
        if (!node) return 0;
        
        curr = (curr << 1) | node.val;
        
        if (!node.left && !node.right) {
            return curr;
        }
        
        return dfs(node.left, curr) + dfs(node.right, curr);
    };
    
    return dfs(root, 0);
};

File: problem/Sum_Root_Leaf/sumRootToLeaf.c
int helper(struct TreeNode *node, int curr)
{
    if (!node)
        return 0;

    curr = (curr << 1) | node->val;

    if (!node->left && !node->right)
    {
        return curr;
    }

    return helper(node->left, curr) + helper(node->right, curr);
}

int sumRootToLeaf(struct TreeNode *root)
{
    return helper(root, 0);
}